### PR TITLE
feat(perf): improve performance of GlamorousComponent#render

### PR DESCRIPTION
**Pull request by kentcdodds**
**What**: This takes some of the ideas from @developit to make the glamorous component render method faster

**Why**: Because that's some hot code!

**How**: copy/paste and blindly looking for other things I figured could be faster...

Ref #43

### CODE REVIEW

File: rollup.config.js [hunk 1]
```diff
@@ -3,22 +3,30 @@ import commonjs from 'rollup-plugin-commonjs'
 import nodeResolve from 'rollup-plugin-node-resolve'
```
Comment (kentcdodds) [inline rollup.config.js]: Pretty much all the changes in this file are to make sure that `process.env.NODE_ENV === 'production'` gets swapped only in the UMD builds.

File: src/__tests__/index.js [hunk 2]
```diff
@@ -213,6 +213,41 @@ test('renders a component with theme properties', () => {
   ).toMatchSnapshotWithGlamor()
 })
 
+test('in development mode the theme is frozen and cannot be changed', () => {
```
Comment (kentcdodds) [inline src/__tests__/index.js]: This is technically a breaking change, but manipulation was never truly supported (or useful) so I'm going to let this one slide.

File: src/index.js [hunk 3]
```diff
@@ -72,30 +72,37 @@ function glamorous(comp, {rootEl, displayName, forwardProps = []} = {}) {
...
 
       render() {
-        const {className, innerRef, ...rest} = this.props
-        const {toForward, cssOverrides} = splitProps(rest, GlamorousComponent)
+        // in this function, we're willing to sacrafice a little on
+        // readability to get better performance because it actually
+        // matters.
+        const props = this.props
+        const {toForward, cssOverrides} = splitProps(props, GlamorousComponent)
+        // freeze the theme object in dev mode
```
Comment (kentcdodds) [inline src/index.js]: Just a handy thing for dev mode so we don't have to make a copy of the object in production mode.

File: src/index.js [hunk 4]
```diff
@@ -72,30 +72,37 @@ function glamorous(comp, {rootEl, displayName, forwardProps = []} = {}) {
...
           glamorStyles: parentGlamorStyles,
           glamorlessClassName,
-        } = extractGlamorStyles(className)
+        } = extractGlamorStyles(props.className)
         const glamorClassName = css(
           ...mappedArgs,
           ...parentGlamorStyles,
           cssOverrides,
         ).toString()
-        const fullClassName = joinClasses(glamorlessClassName, glamorClassName)
```
Comment (kentcdodds) [inline src/index.js]: I figured that calling out to a function that does all the stuff `joinClasses` did was unnecessary weight when the same could be done with simple string concat and `trim`.

**Comment by paulmolluzzo** [comment]
What is the best way to get a performance evaluation that compares this PR to the way it was before?

**Comment by developit** [comment]
@paulmolluzzo ES Bench CI, but it doesn't exist yet (sorry)

**Pull request merged**